src/combinatorics/additive/behrend.lean

```diff
@@ -0,0 +1,192 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Bhavik Mehta
+-/
+import analysis.inner_product_space.pi_L2
+import combinatorics.additive.salem_spencer
+
+/-!
+# Behrend's bound on Roth numbers
+
+This file proves Behrend's lower bound on Roth numbers. This says that we can find a subset of
+`{1, ..., n}` of size `n / exp (O (sqrt (log n)))` which does not contain arithmetic progressions of
+length `3`.
+
+The idea is that the sphere (in the `n` dimensional Euclidean space) doesn't contain arithmetic
+progressions (literally) because the corresponding ball is strictly convex. Thus we can take
+integer points on that sphere and map them onto `ℕ` in a way that preserves arithmetic progressions
+(`behrend.map`).
+
+## Main declarations
+
+* `behrend.sphere`: The intersection of the Euclidean with the positive integer quadrant. This is
+  the set that we will map on `ℕ`.
+* `behrend.map`: The map from the Euclidean space to `n`. It reads off the coordinates as digits in
+  base `d`.
+* `behrend.card_sphere_le_roth_number_nat`: Implicit lower bound on Roth numbers in terms of
+  `behrend.sphere`.
+
+## References
+
+* [Bryan Gillespie, *Behrend’s Construction*]
+  (http://www.epsilonsmall.com/resources/behrends-construction/behrend.pdf)
+* Behrend, F. A., "On sets of integers which contain no three terms in arithmetical progression"
+* [Wikipedia, *Salem-Spencer set*](https://en.wikipedia.org/wiki/Salem–Spencer_set)
+
+## Tags
+
+Salem-Spencer, Behrend construction, arithmetic progression, sphere, strictly convex
+-/
+
+open finset nat real
+open_locale big_operators pointwise
+
+namespace behrend
+variables {α β : Type*} {n d k N : ℕ} {x : fin n → ℕ}
+
+/-!
+### Turning the sphere into a Salem-Spencer set
+
+We define `behrend.sphere`, the intersection of the $$L^2$$ sphere with the positive quadrant of
+integer points. Because the $$L^2$$ closed ball is strictly convex, the $$L^2$$ sphere and
+`behrend.sphere` are Salem-Spencer (`add_salem_spencer_sphere`). Then we can turn this set in
+`fin n → ℕ` into a set in `ℕ` using `behrend.map`, which preserves `add_salem_spencer` because it is
+an additive monoid homomorphism.
+-/
+
+/-- The box `{0, ..., d - 1}^n` as a finset. -/
+def box (n d : ℕ) : finset (fin n → ℕ) := fintype.pi_finset $ λ _, range d
+
+lemma mem_box : x ∈ box n d ↔ ∀ i, x i < d := by simp only [box, fintype.mem_pi_finset, mem_range]
+
+@[simp] lemma card_box : (box n d).card = d ^ n := by simp [box]
+
+/-- The intersection of the sphere of radius `sqrt k` with the integer points in the positive
+quadrant. -/
+def sphere (n d k : ℕ) : finset (fin n → ℕ) := (box n d).filter $ λ x, ∑ i, x i^2 = k
+
+lemma sphere_zero : sphere n d 0 ⊆ 0 :=
+λ x, by simp [sphere, function.funext_iff] {contextual := tt}
+
+lemma sphere_subset_box : sphere n d k ⊆ box n d := filter_subset _ _
+
+lemma norm_of_mem_sphere {n d k : ℕ} {x : fin n → ℕ} (hx : x ∈ sphere n d k) :
```

Comment:
Can you add the link to `metric.sphere` too?

Reply:
What do you have in mind?

Reply:
That one is a subset of the other.

Reply:
```lean
lemma sphere_subset_metric_sphere {n d k : ℕ} :
  (λ x : fin n → ℕ, (pi_Lp.equiv 2 _).symm (coe ∘ x : fin n → ℝ)) '' (sphere n d k : set (fin n → ℕ))
    ⊆ metric.sphere (0 : pi_Lp 2 (λ _ : fin n, ℝ)) (sqrt k) :=
begin
  rintros _ ⟨x, hx, rfl⟩,
  rw mem_sphere_zero_iff_norm,
  exact norm_of_mem_sphere hx
end
```